hal/renesas-rx.c

```diff
@@ -0,0 +1,176 @@
+/* remesas-rx.c
+ *
+ * Stubs for custom HAL implementation. Defines the 
+ * functions used by wolfboot for a specific target.
+ *
+ * Copyright (C) 2022 wolfSSL Inc.
+ *
+ * This file is part of wolfBoot.
+ *
+ * wolfBoot is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * wolfBoot is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <target.h>
+#include "image.h"
+#include "hal.h"
+#include "r_flash_rx.h"
+
+static inline void hal_panic(void)
+{
+    while(1)
+        ;
+}
+
+
+void hal_init(void)
+{
+    if(R_FLASH_Open() != FLASH_SUCCESS)
+        hal_panic();
+
+}
+
+#ifdef __WOLFBOOT
+void hal_prepare_boot(void)
+{
+
+}
+
+#pragma inline_asm longJump
+static void longJump(unsigned long app_offset)
+{
+    jmp   r1;
+}
+
+void do_boot(const uint32_t *app_offset)
+{
+    longJump(app_offset);
+}
+#endif
```

Comment:
please do not redefine `do_boot`. You should link `boot_arm.o` instead. The function does more than just jumping to the app.
In particular, your version of `do_boot` does not:
 - set the stack pointer for the app/OS
 - change the address of the int vector for the app via VTOR (interrupt won't work in the OS you are booting)
 - set up the MPU

Reply:
> please do not redefine `do_boot`. You should link `boot_arm.o` instead. The function does more than just jumping to the app. In particular, your version of `do_boot` does not:
> 
> * set the stack pointer for the app/OS
> * change the address of the int vector for the app via VTOR (interrupt won't work in the OS you are booting)
> * set up the MPU

The demo application includes BSP startup which takes care changing SP, interruption vector and other settings.
I am assuming you can simply jump to the entry point.

RX is not ARM architecture, but Renesas proprietary. Would it be good to have it in something like boot_RenesasRx.c?

Reply:
Sorry, my bad for missing the architecture of this port.

`do_boot` is part of the architecture code, not target. Yes in general it would make sense to have a boot_$(ARCH).c like all other architectures, but only worth it if we plan to port to other targets in the same $ARCH.

Usually the bring-up code is self-contained in wolfBoot. We define our own interrupt vector, which is empty apart for the critical faults, because we want all interrupt to be off during the bootloader stage. We usually write our own reset ISR to have the flow under control.

Relocation of the interrupt vector is also necessary before staging. Your simple application works even without setting a new offset, but if you loaded an OS or a demo that uses interrupts in your application code it won't probably trigger any interrupts because the offset has not been set. I suggest to do the same we do with other architectures to avoid this problem. The register equivalent to ARM `VTOR` seems to be `EXTB` in Renesas-RX.

Reply:
Sorry, I have pushed boot_renesas_rx.c for do_boot before noticing your comment. I would arrange it either way you feel appropriate.

Regarding wolfBoot and the application configuration, I assume I am doing the way as your comment. 
wolfBoot and the sample application are totally independent, self-contained. Each has own interrupt vector appropriately located. e2Studio for Rx takes care the location and includes own startup routine for each to setup ISP, USP or IV, as long as you have a project for each. This looks like working, so far, and will see if it works for more complicated configurations.

Based on this configuration, I assume, wolfBoot can run totally interrupt disabled or its own very limited IV.